measurements/pages/__init__.py

```diff
@@ -204,34 +208,69 @@ def files_in_country(country_code):
-        # XXX
-        # We have duplicate measurements :(
-        # So the below exception actually happens. This should be resolved
-        # in the data processing pipeline.
-        #   Report.filename == filename).one()
-    except NoResultFound:
-        raise NotFound("No file with that filename found")
-    except MultipleResultsFound:
-        # This should never happen.
-        raise HTTPException("Duplicate measurement detected")
-
-    if report is None:
+def decompress_autoclaved(autoclaved_filename, frames_off, frames_size, intra_off, intra_size):
+    def generator():
+        url = urljoin(current_app.config['AUTOCLAVED_BASE_URL'], autoclaved_filename)
+
+        current_app.logger.debug("Fetching autoclaved from: %s" % url)
+        current_app.logger.debug("  frames_off: %d" % frames_off)
+        current_app.logger.debug("  frames_size: %d" % frames_size)
+        current_app.logger.debug("  intra_off: %d" % intra_off)
+        current_app.logger.debug("  intra_size: %d" % intra_size)
+        try:
+            r = requests.get(url, headers={"Range": "bytes={}-{}".format(frames_off, frames_off + frames_size)},
+                                stream=True)
+            streamed_data = 0
+            for chunk in lz4framed.Decompressor(r.raw):
+                current_app.logger.debug("chunking away 1")
+                d = chunk
+                current_app.logger.debug("chunking away 2")
+                if streamed_data == 0 and intra_off > 0:
+                    d = chunk[intra_off:]
+                if streamed_data > (intra_size + len(d)):
+                    d = chunk[:(streamed_data - intra_size)]
+                yield d
+                streamed_data += len(d)
+            current_app.logger.debug("sent: %d" % streamed_data)
+        except Exception as exc:
+            current_app.logger.error("failed to fetch streamed data: %s" % exc)
+            raise HTTPException("Failed to fetch data")
+    return generator
+
+@pages_blueprint.route('/files/download/<path:textname>')
+def files_download(textname):
+    q = current_app.db_session.query(
+            Measurement.id.label('m_id'),
+            Measurement.report_no.label('report_no'),
+            Measurement.frame_off.label('frame_off'),
+            Measurement.frame_size.label('frame_size'),
+            Measurement.intra_off.label('intra_off'),
+            Measurement.intra_size.label('intra_size'),
+            Report.report_no.label('r_report_no'),
+            Report.autoclaved_no.label('r_autoclaved_no'),
+            Autoclaved.filename.label('a_filename'),
+            Autoclaved.autoclaved_no.label('a_autoclaved_no'),
+    ).filter(Report.textname == textname) \
+        .join(Report, Report.report_no == Measurement.report_no) \
+        .join(Autoclaved, Autoclaved.autoclaved_no == Report.autoclaved_no) \
+        .order_by(Measurement.frame_off.asc(), Measurement.intra_off.asc())
+
+    msmts = q.all()
```

Comment:
Is there a reason to transfer all the rows form DB to application? Is it size accounting?

Reply:
I couldn't come up with a better way of obtaining the first and last row without doing 2 queries. It seems slightly more efficient to waste some bandwidth rather than wasting CPU cycles doing two queries.

Reply:
The question is not wasting CPU **OR** bandwidth. I think, that the most time-consuming part of "bandwidth" consuming process is deserialization of data to python structures. I'm not motivated enough to test that, but here is some query if you want to :-)

```sql
SELECT * FROM (
  SELECT
    filename, frame_off, frame_size, intra_off, intra_size,
    row_number() OVER (ORDER BY frame_off, intra_off) AS rn,
    count(*) OVER () AS total_count,
    sum(intra_size + 1) OVER () AS report_size
  FROM autoclaved JOIN report USING (autoclaved_no) JOIN measurement USING (report_no)
  WHERE textname = '2017-05-17/20170516T205044Z-FR-AS34572-web_connectivity-20170516T205043Z_AS34572_HINf1L4OfIdyHkaDQ6WFiYaPkuIpHYeWRkigik4X66YueNOtGt-0.2.0-probe.json'
) t1 WHERE rn = 1 OR rn = total_count;
```

The tricky part is cutting the tail precisely. As far as I see, there are two

Reply:
Well that would be CPU "wasted" on the side of the measurements API not on the database, which is what I am actually most concerned by.

In any case this performs adequately and works, so I don't have plans on changing it at the moment until there is evidence that another working approach yields a significant performance increase.

Reply:
> significant performance increase

There is bad case having ~10k measurements within single report file, `2014-11-22/20141122T040940Z-US-AS1968-tcp_connect-no_report_id-0.1.0-probe.yaml`. Please, check if it has acceptable performance.

Reply:
@darkk:

> The tricky part is cutting the tail precisely. I assume that all report files have measurements delimited by single \n (that's why intra_size + 1)

What if we silently break this assumption by mistake? That is, say on Windows I make a programming mistake in MK such that I write `\n` that the standard library converts to `\r\n` because I forgot about binary format?

> There is bad case having ~1'000'000 measurements within single report file

Fair point. I wonder how many files are similar to that. Do you have a statistic? If that is fairly common, we probably have a problem. (Is it possible to tweak the pipeline to avoid producing these corner cases, or it is intrinsic and my question does not make sense because I don't understand the pipeline?)

Reply:
> What if we silently break this assumption by mistake.

The very best way is to store some checksum in database and verify it on the fly. But there is no alike checksum right now, I've not stored it as this checksum is not "immutable" and may vary if we re-sanitise private files. I think that acceptable way is to verify that served file starts with `{` and ends with `}\n`. IMHO, that's sufficient basic protection.

> I wonder how many files are similar to that.

That's the biggest one, but it should probably be still downloadable and should not lead to DoS. The second prize is for report with ~70'000 measurements.

Reply:
Ok, I ended up implementing your method in the latest commit. I tested it on a couple of sample reports, but not that thoroughly.

In my limited testing I didn't actually notice a considerable performance boost, but I also didn't run it against the 1M measurement entries.

I actually didn't know there were such reports (I my limited searching I only found some 10k+ reports), so I do agree that in those case it makes sense to not waste all that amount of bandwidth and CPU on the ooni-measurements side.

Reply:
> found some 10k+ reports

I just run `select report_no, count(*) as cntmsm from measurement group by report_no order by cntmsm desc limit 5;` against hkgmetadb.